Revamp `KeychainTxOutIndex` API to be safer

evanlinjin opened the pull request:
Closes #1268

### Description

Previously `SpkTxOutIndex` methods can be called from `KeychainTxOutIndex` due to the `DeRef` implementation. However, the internal `SpkTxOut` will also contain lookahead spks resulting in an error-prone API.

`SpkTxOutIndex` methods are now not directly callable from `KeychainTxOutIndex`. Methods of `KeychainTxOutIndex` are renamed for clarity. I.e. methods that return an unbounded spk iter are prefixed with `unbounded`.

In addition to this, I also optimized the peek-address logic of `bdk::Wallet` using the optimized `<SpkIterator as Iterator>::nth` implementation.

### Notes to the reviewers

This is mostly refactoring, but can also be considered a bug-fix (as the API before was very problematic).

### Changelog notice

Changed
* Wallet's peek-address logic is optimized by making use of `<SpkIterator as Iterator>::nth`.
* `KeychainTxOutIndex` API is refactored to better differentiate between methods that return unbounded vs stored spks.
* `KeychainTxOutIndex` no longer directly exposes `SpkTxOutIndex` methods via `DeRef`. This was problematic because `SpkTxOutIndex` also contains lookahead spks which we want to hide.

Added
* `SpkIterator::descriptor` method which gets a reference to the internal descriptor. 

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

~* [ ] I've added tests for the new feature~
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
feat(chain): add `SpkIterator::descriptor` method

Otherwise there will be no way to view the descriptor of the
`SpkIterator`.

evanlinjin committed:
feat(chain): make output of `SpkTxOutIndex::unused_spks` cloneable

This allows us to pass this to chain sources without calling
`Iterator::collect` first.

evanlinjin committed:
refactor(chain)!: revamp `KeychainTxOutIndex` API

Previously `SpkTxOutIndex` methods can be called from
`KeychainTxOutIndex` due to the `DeRef` implementation. However, the
internal `SpkTxOut` will also contain lookahead spks resulting in an
error-prone API.

`SpkTxOutIndex` methods are now not directly callable from
`KeychainTxOutIndex`. Methods of `KeychainTxOutIndex` are renamed for
clarity. I.e. methods that return an unbounded spk iter are prefixed
with `unbounded`.

evanlinjin committed:
fix(wallet): use efficient peek address logic

Changes the peek address logic to use the optimized `Iterator::nth`
implementation of `SpkIterator`.

Additionally, docs are added for panics that will occur when the caller
requests for addresses with out-of-bound derivation indices (BIP32).

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -792,36 +819,36 @@ impl<D> Wallet<D> {
     /// Note carefully that iterators go over **all** script pubkeys on the keychains (not what
     /// script pubkeys the wallet is storing internally).
-    pub fn spks_of_all_keychains(
+    pub fn all_unbounded_spk_iters(
```
The doc for this method is missing a \`, can you fix it please? Line 814, there's no ` at the start of the word "External"

danielabrozzoni commented on example-crates/example_esplora/src/main.rs:
```diff
@@ -235,32 +235,32 @@ fn main() -> anyhow::Result<()> {
-                        .all_spks()
-                        .iter()
-                        .map(|(k, v)| (*k, v.clone()))
+                        .revealed_spks()
```
Why is this iterating over revealed only? Was using `all_spks()` a bug?

evanlinjin replied:
Yes `all_spks` includes lookahead spks because it if `Deref`ed from `SpkTxOutIndex`. We want to hide all lookahead spks.

danielabrozzoni replied:
Should we clarify this change in the changelog?

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -792,36 +819,36 @@ impl<D> Wallet<D> {
-    /// [`spks_of_all_keychains`]: Self::spks_of_all_keychains
-    pub fn spks_of_keychain(
+    /// [`all_unbounded_spk_iters`]: Self::all_unbounded_spk_iters
+    pub fn unbounded_spk_iter(
```
Is "unbounded" meant to be the contrary of "stored only"? 

To me, "unbounded_spk_iter" makes it sound like that there is a way to get a "bounded" spk iter, but then there's no method called like that. Maybe "all" would be a better word to communicate "it's all of the spks, not the stored only" - but then I wouldn't know how to call "all_unbounded_spk_iters" 😅 

This is definitely not a blocker, but I'm not convinced that "unbounded" is the right word here

evanlinjin replied:
I don't think `all` is a better option. It can be interpreted as "all-revealed" or "all-stored". I'll be open for a better name, let me know if you have any further ideas.

LLFourn replied:
I like unbounded better. Although there is no "bounded" version it implies the others are bounded (which they are). "unbounded" tells the caller that it's never going to stop so it would only make sense to use this in a "stop gap" context.

danielabrozzoni commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -133,8 +125,11 @@ impl<K> KeychainTxOutIndex<K> {
 impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     /// Return a reference to the internal [`SpkTxOutIndex`].
+    ///
+    /// **WARNING:** The internal index will contain lookahead spks.
```
nit: maybe we should expand a little bit on stored vs revealed vs lookahead.

LLFourn replied:
I think we should do this on the main type. It already does this a bit but it could be clearer.

evanlinjin replied:
83e7b7ec402bef27c3f83876f346ec2abd23eff1

danielabrozzoni reviewed (commented):
A few comments. I still haven't finished my review, but I wanted to post this in the meantime :)

LLFourn commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -144,8 +139,103 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+    }
+
+    /// Return a reference to the internal map of keychain to descriptors.
+    pub fn keychain_descriptors(&self) -> &BTreeMap<K, Descriptor<DescriptorPublicKey>> {
```
I don't think we need `keychains` and `keychain_descriptors`.

evanlinjin committed:
docs(wallet): improve docs for unbounded spk iterator methods

evanlinjin committed:
docs(chain): improve `KeychainTxOutIndex` docs

More clarity for revealed/lookahead spks and give usecases.

LLFourn reviewed (approved):
ACK 83e7b7ec402bef27c3f83876f346ec2abd23eff1

danielabrozzoni commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -133,8 +165,12 @@ impl<K> KeychainTxOutIndex<K> {
     /// Return a reference to the internal [`SpkTxOutIndex`].
+    ///
+    /// **WARNING:** The internal index will contain lookahead spks. Refer to
+    /// [struct-level docs](KeychainTxOutIndex) for more about `lookahead`.
```
Do you think we will eventually avoid exposing `inner`, or is it useful for some use cases, and so exposing it with a warning is good enough?

evanlinjin replied:
I would assume in some cases getting lookahead spks will be handy.

I.e. we haven't finalized how a CBF chain-source will look like and how it will interact with `bdk_chain` structures. However, it might make sense to match lookahead spks against GCS (filters). Or maybe using an *unbounded* spk iterator makes more sense?

Maybe later on, we determine that it will make sense to use helper methods on `KeychainTxOutIndex` instead. However, for most usecases now, outputting lookahead spks is almost always non-intentional.

evanlinjin replied:
In summary, I think we should keep `.inner` with a warning (for now).

danielabrozzoni reviewed (commented):
From my understanding, this PR is making some `SpkTxOutIndex` methods slightly more difficult to call (you need to use `KeychainTxOutIndex::inner` instead of just using `Deref`).

These are SpkTxOutIndex methods that are not directly exposed from `KeychainTxOutIndex`:
- `all_spks`
- `insert_spk`
- `outputs_in_range`
- `txouts`
- `txouts_in_tx`

It makes sense to me that `insert_spk` and `all_spks` are not exposed: you should let the `KeychainTxOutIndex` insert the spks, and `all_spks` exposes the lookahead spks.

But I just can't figure out why `outputs_in_range`, `txouts`, and `txouts_in_tx` shouldn't be exposed as well - is there a particular reason?

evanlinjin commented:
> But I just can't figure out why `outputs_in_range`, `txouts`, and `txouts_in_tx` shouldn't be exposed as well - is there a particular reason?

@danielabrozzoni I didn't include them because they weren't used in the examples. I guess it doesn't hurt to include?

evanlinjin committed:
feat(chain): add txout methods to `KeychainTxOutIndex`

`txouts` and `txouts_in_tx` are exposed from `SpkTxOutIndex`, but
modified to remove nested unions.

Add `keychain_outpoints_in_range` that iterates over outpoints of a
given keychain derivation range.

danielabrozzoni reviewed (approved):
ACK 71fff1613d1c167d180636857063e181f6acedf1
